Enhance OpenTelemetry span logging to include query settings

sharathks118 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Enhance OpenTelemetry span logging to include query settings

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

sharathks118 committed:
Enhance OpenTelemetry span logging to include query settings

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=70011) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=70011) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=70011) it.</sub>

kssenii commented on tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:
```diff
@@ -72,6 +85,7 @@ check_query_span "$query_id"
 query_id=$(${CLICKHOUSE_CLIENT} -q "select generateUUIDv4()");
 execute_query $query_id 'select * from opentelemetry_test format Null'
 check_query_span $query_id
+check_query_settings
```
need to also update according `tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.reference` file

sharathks118 replied:
Update test script and reference file as per review comments

kssenii commented on programs/server/config.xml:
```diff
@@ -202,6 +202,10 @@
     <interserver_http_port>9009</interserver_http_port>
 
+    <settings>
+        <log_query_settings>1</log_query_settings>
```
This is not the correct place to modify settings for the tests, please modify ClickHouse/tests/configs/ instead

sharathks118 replied:
Addressed the review comments

kssenii commented on programs/server/users.xml:
```diff
@@ -5,6 +5,8 @@
     <profiles>
         <!-- Default settings. -->
         <default>
+            <allow_introspection_functions>1</allow_introspection_functions>
```
This does not look required for the test. Anyway same comment as above that this is not the correct place to modify settings for tests.

kssenii commented on tests/queries/0_stateless/02421_simple_queries_for_opentelemetry.sh:
```diff
@@ -44,6 +44,19 @@ ${CLICKHOUSE_CLIENT} -q "
+{
+    ${CLICKHOUSE_CLIENT} -q "
+        SYSTEM FLUSH LOGS;
+        SELECT attribute
```
This test will give too big output, better to search for specific query settings.

sharathks118 replied:
Update test script and reference file as per review comments

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #70011 from sharathks118/ClickHouse:setting-opentelemetry -->
*This is an automated comment for commit 9f8964a547b5219a580f8b64655b56e2bfa70f61 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70011/9f8964a547b5219a580f8b64655b56e2bfa70f61/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

sharathks118 committed:
Fixed review comments

Removed config files changes

sharathks118 committed:
Fix indentation in shell script

kssenii commented:
@sharathks118 Hi, there is an issue with checks.
For example, `Fast test — Cannot build ClickHouse `.
```
2024-10-01 14:02:09 /repo/src/Interpreters/executeQuery.cpp:573:22: error: no member named 'log_query_settings' in 'DB::Settings'
2024-10-01 14:02:09   573 |         if (settings.log_query_settings)
2024-10-01 14:02:09       |             ~~~~~~~~ ^
2024-10-01 14:02:09 /repo/src/Interpreters/executeQuery.cpp:575:50: error: no member named 'allChanged' in 'DB::Settings'
2024-10-01 14:02:09   575 |             for (const auto & setting : settings.allChanged())
2024-10-01 14:02:09       |                                         ~~~~~~~~ ^
2024-10-01 14:02:09 2 errors generated.
```

sharathks118 committed:
Handle fast test failure

FrankChen021 commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -563,6 +563,46 @@ void logQueryFinish(
+                        break;
+                }
+
+                query_span->addAttribute(name, value_str);
```
I would like to suggest adding a prefix(like `clickhouse.setting.`) to the name so that all these query setting names in the span log can be easily distinguished from other attributes.

sharathks118 replied:
Addressed review comments by adding a "clickhouse.setting." prefix to query setting names

FrankChen021 commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -563,6 +563,46 @@ void logQueryFinish(
+            query_span->addAttribute("clickhouse.user", user_name);
+        }
+
+        if (settings[Setting::log_query_settings])
```
If this feature is controlled by this setting, it's better to update docs and the setting description in the code to reflect this change.

sharathks118 replied:
Updated documentation and setting descriptions

sharathks118 committed:
Add prefix "clickhouse.setting." to query setting names in span log
Updated relevant documentation and setting descriptions to reflect this change.

sharathks118 committed:
Merge branch 'master' into setting-opentelemetry

sharathks118 committed:
Updated log_query_settings description in Settings.cpp

sharathks118 commented:
The test failure in the pipeline is not related to my changes.

kssenii commented on docs/en/operations/settings/settings.md:
```diff
@@ -540,6 +540,13 @@ Default value: 0
 
 Allow non-const timezone arguments in certain time-related functions like toTimeZone(), fromUnixTimestamp*(), snowflakeToDateTime*()
 
+## log_query_settings {#log-query-settings}
+Type: Bool
```
```suggestion
## log_query_settings {#log-query-settings}

Type: Bool
```

kssenii commented on docs/en/operations/opentelemetry.md:
```diff
@@ -31,6 +31,9 @@ The table must be enabled in the server configuration, see the `opentelemetry_sp
 
 The tags or attributes are saved as two parallel arrays, containing the keys and values. Use [ARRAY JOIN](../sql-reference/statements/select/array-join.md) to work with them.
 
+## Log-query-settings
+ClickHouse allows you to log changes to query settings during query execution. When enabled, any modifications made to query settings will be recorded in the OpenTelemetry span log. This feature is particularly useful in production environments for tracking configuration changes that may affect query performance.
```
```suggestion
## Log-query-settings

ClickHouse allows you to log changes to query settings during query execution. When enabled, any modifications made to query settings will be recorded in the OpenTelemetry span log. This feature is particularly useful in production environments for tracking configuration changes that may affect query performance.
```

kssenii committed:
Update docs/en/operations/settings/settings.md

kssenii committed:
Update docs/en/operations/opentelemetry.md

kssenii commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -564,6 +564,47 @@ void logQueryFinish(
+                Field value = settings.get(name);
+                String value_str;
+
+                switch (value.getType())
```
Let's use here `convertFieldToString` from `Common/FieldVisitorToString.h`

sharathks118 replied:
updated the code to use convertFieldToString from Common/FieldVisitorToString.h

sharathks118 committed:
Refactor executeQuery.cpp to use convertFieldToString

Algunenano commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -564,6 +565,25 @@ void logQueryFinish(
+
+        if (settings[Setting::log_query_settings])
+        {
+            auto changed_settings_names = settings.getChangedNames();
```
You already have this in settings.changes(). No need to add a new function and call .get() every time

rschu1ze commented on docs/en/operations/opentelemetry.md:
```diff
@@ -31,6 +31,10 @@ The table must be enabled in the server configuration, see the `opentelemetry_sp
 
 The tags or attributes are saved as two parallel arrays, containing the keys and values. Use [ARRAY JOIN](../sql-reference/statements/select/array-join.md) to work with them.
 
+## Log-query-settings
```
@sharathks118 Without a link to the corresponding setting in "settings.md", readers will be confused by "Log-query-settings" and below description.
